src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp

```diff
@@ -775,56 +447,39 @@ void ShenandoahPrepareForCompactionTask::work(uint worker_id) {
   GrowableArray<ShenandoahHeapRegion*> empty_regions((int)_heap->num_regions());
 
   if (_heap->mode()->is_generational()) {
-    ShenandoahHeapRegion* old_to_region = (from_region->is_old())? from_region: nullptr;
-    ShenandoahHeapRegion* young_to_region = (from_region->is_young())? from_region: nullptr;
     ShenandoahPrepareForGenerationalCompactionObjectClosure cl(_preserved_marks->get(worker_id),
-                                                               empty_regions,
-                                                               old_to_region, young_to_region,
-                                                               worker_id);
-    while (from_region != nullptr) {
-      assert(is_candidate_region(from_region), "Sanity");
-      log_debug(gc)("Worker %u compacting %s Region " SIZE_FORMAT " which had used " SIZE_FORMAT " and %s live",
-                    worker_id, from_region->affiliation_name(),
-                    from_region->index(), from_region->used(), from_region->has_live()? "has": "does not have");
-      cl.set_from_region(from_region);
-      if (from_region->has_live()) {
-        _heap->marked_object_iterate(from_region, &cl);
-      }
-      // Compacted the region to somewhere else? From-region is empty then.
-      if (!cl.is_compact_same_region()) {
-        empty_regions.append(from_region);
-      }
-      from_region = it.next();
-    }
-    cl.finish();
-
-    // Mark all remaining regions as empty
-    for (int pos = cl.empty_regions_pos(); pos < empty_regions.length(); ++pos) {
-      ShenandoahHeapRegion* r = empty_regions.at(pos);
-      r->set_new_top(r->bottom());
-    }
+                                                               empty_regions, from_region, worker_id);
+    prepare_for_compaction(cl, empty_regions, it, from_region);
   } else {
     ShenandoahPrepareForCompactionObjectClosure cl(_preserved_marks->get(worker_id), empty_regions, from_region);
-    while (from_region != nullptr) {
-      assert(is_candidate_region(from_region), "Sanity");
-      cl.set_from_region(from_region);
-      if (from_region->has_live()) {
-        _heap->marked_object_iterate(from_region, &cl);
-      }
+    prepare_for_compaction(cl, empty_regions, it, from_region);
+  }
+}
 
-      // Compacted the region to somewhere else? From-region is empty then.
-      if (!cl.is_compact_same_region()) {
-        empty_regions.append(from_region);
-      }
-      from_region = it.next();
+template<typename ClosureType>
+void ShenandoahPrepareForCompactionTask::prepare_for_compaction(ClosureType& cl,
```

Comment:
Looking at this method, it looks like it actually belongs as a public method in the closure, with that closure's methods invoked here becoming private to the closure. Makes for a narrower public API all around and keeps the loop in the right place.

Reply:
Yes, I think you are right. When I was writing that comment, I was thinking that the loop could just be duplicated into each of the two closures. I guess may be that doesn't particularly make it any simpler and moreover leads to some avoidable duplication. I'll leave it up to you as to which makes more sense to you.

Reviewed and approved!

Reply:
I'll leave it for now. This form also reduces some of the delta with upstream (that is, with fewer changes upstream).